File: peerreviewplugin/0.9/Code/codereview-plugin/codereview/peerReviewMain.py
#Copyright (C) 2006 Anthony Panozzo
#Copyright (C) 2006 Adam Westhusing
#All rights reserved. 

#This file is part of The Trac Peer Review Plugin

#The Trac Peer Review Plugin is free software; you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation; either version 2 of the License, or
#(at your option) any later version.

#The Trac Peer Review Plugin is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with The Trac Peer Review Plugin; if not, write to the Free Software
#Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

# Provides functionality for main page
# Works with peerReviewMain.cs

from trac.core import *
from trac.web.chrome import INavigationContributor, ITemplateProvider
from trac.web.main import IRequestHandler
from trac.perm import IPermissionRequestor
from trac import util
from trac.util import escape, Markup
from codereview.dbBackend import *
from trac.web.chrome import add_stylesheet


class UserbaseModule(Component):
    implements(INavigationContributor, IRequestHandler, ITemplateProvider, IPermissionRequestor)
        
    # INavigationContributor methods
    def get_active_navigation_item(self, req):
        return 'peerReviewMain'
                
    def get_navigation_items(self, req):
        if not (req.perm.has_permission('CODE_REVIEW_DEV') or req.perm.has_permission('CODE_REVIEW_MGR')):
            return
        yield ('mainnav', 'peerReviewMain',
               Markup('<a href="%s">Peer Review</a>',
                      self.env.href.peerReviewMain()))
        
    # IRequestHandler methods
    def match_request(self, req):
        return req.path_info == '/peerReviewMain'


    # IPermissionRequestor methods
    def get_permission_actions(self):
         return ['CODE_REVIEW_DEV', 'CODE_REVIEW_MGR']
                                        

    def process_request(self, req):
        # test whether this user is a manager or not
        if req.perm.has_permission('CODE_REVIEW_MGR'):
            req.hdf['author'] = "manager"
            req.hdf['manager'] = 1
        else:
            req.perm.assert_permission('CODE_REVIEW_DEV')
            req.hdf['author'] = "notmanager"
            req.hdf['manager'] = 0

        # set up dynamic links
        req.hdf['trac.href.peerReviewMain'] = self.env.href.peerReviewMain()
        req.hdf['trac.href.peerReviewNew'] = self.env.href.peerReviewNew()
        req.hdf['trac.href.peerReviewSearch'] = self.env.href.peerReviewSearch()
        req.hdf['trac.href.peerReviewOptions'] = self.env.href.peerReviewOptions()

        req.hdf['main'] = "yes"
        req.hdf['create'] = "no"
        req.hdf['search'] = "no"
        req.hdf['options'] = "no"

        req.hdf['trac.href.peerReviewView'] = self.env.href.peerReviewView()
        req.hdf['username'] = util.get_reporter_id(req)

        db = self.env.get_db_cnx()
        codeReview = CodeReviewStruct(None)
        dbBack = dbBackend(db)
        codeReviewArray = dbBack.getMyCodeReviews(util.get_reporter_id(req))
        assignedReviewArray = dbBack.getCodeReviews(util.get_reporter_id(req))
        managerReviewArray = dbBack.getCodeReviewsByStatus("Ready for inclusion")
        reviewReturnArray = []
        assignedReturnArray = []
        managerReturnArray = []
        dataArray = []

        # fill the table of currently open reviews
        for struct in codeReviewArray:
            if struct.Status != "Closed":
                dataArray.append(struct.IDReview)
                dataArray.append(struct.Author)
                dataArray.append(struct.Status)
                dataArray.append(util.format_date(struct.DateCreate))
                dataArray.append(struct.Name)
                reviewReturnArray.append(dataArray)
                dataArray = []
                dataArray = []
        
        # fill the table of code reviews currently assigned to you
        for struct in assignedReviewArray:
            if struct.Status != "Closed":
                dataArray.append(struct.IDReview)
                dataArray.append(struct.Author)
                dataArray.append(struct.Name)
                dataArray.append(util.format_date(struct.DateCreate))            
                reviewstruct = dbBack.getReviewerEntry(struct.IDReview, util.get_reporter_id(req))
                if reviewstruct.Vote == "-1":
                    dataArray.append('Not voted')
                elif reviewstruct.Vote == "0":
                    dataArray.append('Rejected')
                elif reviewstruct.Vote == "1":
                    dataArray.append('Accepted')
                assignedReturnArray.append(dataArray)
                dataArray = []
                dataArray = []

        # fill the table of reviews assigned to you in a manager role
        for struct in managerReviewArray:
            if struct.Status != "Closed":
                dataArray.append(struct.IDReview)
                dataArray.append(struct.Author)
                dataArray.append(struct.Name)
                dataArray.append(util.format_date(struct.DateCreate))
                managerReturnArray.append(dataArray)
                dataArray = []

        req.hdf['reviewReturnArrayLength'] = len(reviewReturnArray)
        req.hdf['assignedReturnArrayLength'] = len(assignedReturnArray)
        req.hdf['managerReviewArrayLength'] = len(managerReviewArray)

        req.hdf['myCodeReviews'] = reviewReturnArray
        req.hdf['assignedReviews'] = assignedReturnArray
        req.hdf['managerReviews'] = managerReturnArray
        add_stylesheet(req, 'common/css/code.css')
        add_stylesheet(req, 'common/css/browser.css')	
        return 'peerReviewMain.cs', None
                
    # ITemplateProvider methods
    def get_templates_dirs(self):
        """
        Return the absolute path of the directory containing the provided
        ClearSilver templates.
        """
        from pkg_resources import resource_filename
        return [resource_filename(__name__, 'templates')]

    def get_htdocs_dirs(self):
        from pkg_resources import resource_filename
        return [('hw', resource_filename(__name__, 'htdocs'))]

File: peerreviewplugin/0.10/codereview/templates/peerReviewMain.cs
<?cs include "header.cs" ?>
<?cs include "macros.cs" ?>

<!-- Top navigation bar -->
<div id="ctxtnav" class="nav"> 
	<h2>Peer Review Navigation</h2> 
	<ul>
		<li class="first">
			<?cs if:main == "yes" ?>My Code Reviews
			<?cs else ?><a href="<?cs var:trac.href.peerReviewMain ?>">My Code Reviews</a>
			<?cs /if ?>
		</li>
		<li>
			<?cs if:create == "yes" ?>Create a Code Review
			<?cs else ?><a href="<?cs var:trac.href.peerReviewNew ?>">Create a Code Review</a>
			<?cs /if ?>
		</li>
		<li <?cs if:manager == 0 ?>class="last"<?cs /if ?>>
			<?cs if:search == "yes" ?>Search Code Reviews
			<?cs else ?><a href="<?cs var:trac.href.peerReviewSearch ?>">Search Code Reviews</a>
			<?cs /if ?>
		</li>
		<?cs if:manager == 1 ?>
		<li class="last">
			<?cs if:option == "yes" ?>Manager Options
			<?cs else ?><a href="<?cs var:trac.href.peerReviewOptions ?>">Manager Options</a>
			<?cs /if ?>
		</li>
		<?cs /if ?>
	</ul> 
</div> 


<p><h1>My Code Reviews</h1></p>

<br>
<!-- If we are a manager, display a table containing all of the code reviews requiring your approval. -->
<?cs if:author == "manager" ?>
<p><h2>This is a list of the code reviews currently requiring your approval.</h2></p>
<table class="listing" id="approvereviewlist">
	<thead>
		<tr>
			<td>Review</td>
			<td>Author</td>
			<td>Name</td>
			<td>Date Opened</td>
		</tr>
	</thead>
	<tbody>
        <!-- if there are no reviews open for you, state this in the table -->
        <?cs if: (managerReviewArrayLength == 0) ?>
			<tr>		
			<td>There are no code reviews requiring your approval.</td>
			<td></td>
			<td></td>
			<td></td>
			</tr>
		<?cs else ?>
			<?cs each:item = managerReviews ?>
				<tr class="<?cs if:name(item) % #2 ?>even<?cs else ?>odd<?cs /if ?>">
                    <!-- Set up the table with the appropriate values and then links to view the review -->
               		<td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[0]?></a></td>
                    <td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[1]?></a></td>
                    <td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[2]?></a></td>
                    <td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[3]?></a></td>
				</tr>
			<?cs /each ?>
		<?cs /if ?>
		
	</tbody>
</table>
<br>
<br>
<?cs /if ?>

<p><h2>This is a list of your currently open code reviews.</h2></p>

<!-- Display your currently open code reviews in a table -->
<table class="listing" id="myreviewlist">
	<thead>
		<tr>
			<td>Review</td>
			<td>Name</td>
			<td>Status</td>
			<td>Date Opened</td>
		</tr>
	</thead>
	<tbody>
		<!-- if there are no reviews open for you, state this in the table -->
		<?cs if:(reviewReturnArrayLength == 0) ?>
			<tr>
			<td>Your have no code reviews currently open.</td>
			<td></td>
			<td></td>
			<td></td>
			</tr>
		<?cs else ?>
			<?cs each:item = myCodeReviews ?>
				<tr class="<?cs if:name(item) % #2 ?>even<?cs else ?>odd<?cs /if ?>">
                    <!-- Set up the table with the appropriate values and then links to view the review -->
                    <td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[0]?></a></td>
                    <td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[4]?></a></td>
                    <td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[2]?></a></td>
                    <td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[3]?></a></td>
				</tr>
			<?cs /each ?>
		<?cs /if ?>
	</tbody>
</table>

<br>
<br>
<p><h2>This is a list of code reviews currently assigned to you.</h2></p>

<!-- Display your currently assigned code reviews in a table -->
<table class="listing" id="assignedreviewlist">
	<thead>
		<tr>
			<td>Review</td>
			<td>Author</td>
			<td>Name</td>
			<td>My Vote</td>
			<td>Date Opened</td>
		</tr>
	</thead>
	<tbody>
        <!-- if there are no reviews open for you, state this in the table -->
        <?cs if: (assignedReturnArrayLength == 0) ?>
			<tr>
			<td>There are no code reviews assigned to you.</td>
			<td></td>
			<td></td>
			<td></td>
			<td></td>
			</tr>
		
		<?cs else ?>
			<?cs each:item = assignedReviews ?>
				<tr class="<?cs if:name(item) % #2 ?>even<?cs else ?>odd<?cs /if ?>">
                    <!-- Set up the table with the appropriate values and then links to view the review -->
                    <td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[0]?></a></td>
                    <td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[1]?></a></td>
                    <td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[2]?></a></td>
                    <td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[4]?></a></td>
                    <td><a href="<?cs var:trac.href.peerReviewView ?>?Review=<?cs var:item[0]?>"><?cs var:item[3]?></a></td>
				</tr>
			<?cs /each ?>
		<?cs /if ?>
	</tbody>
</table>
<br>
<br>

<?cs include "footer.cs" ?>
